transaction/core/src/ring_ct/rct_bulletproofs.rs

```diff
@@ -95,8 +95,82 @@ pub struct SignedInputRing {
     pub signed_digest: Option<[u8; 32]>,
 }
 
+/// A structure that contains all the data required to sign a transaction that
+/// does not rely on having the spend private key.
+/// Exposing this allows implementing various offline/hardware signing
+/// solutions. The idea is that the data in this object can be generated by only
+/// having the view private key for a set of input TxOuts, and it can then be
+/// moved to a separate machine/service that then takes it + the spend private
+/// key and uses that to generate a fully signed transaction.
+#[derive(Clone, Debug, Deserialize, Digestible, Eq, PartialEq, Serialize)]
+pub struct SigningData {
+    /// Transaction extended message.
+    pub extended_message_digest: Vec<u8>,
+
+    /// Pseudo output blindings.
+    pub pseudo_output_blindings: Vec<Scalar>,
+
+    /// Commitments of value equal to each real input.
+    pub pseudo_output_commitments: Vec<CompressedCommitment>,
+
+    /// Proof that all pseudo_outputs and transaction outputs are in [0, 2^64).
+    /// This contains range_proof.to_bytes().
+    pub range_proof_bytes: Vec<u8>,
+
+    /// A range proof, one for each token id that is used in the transaction.
+    ///
+    /// The range proofs correspond to the sorted order of token ids used.
+    ///
+    /// Note: This is EMPTY if mixed transactions is not enabled.
+    pub range_proofs: Vec<Vec<u8>>,
+
+    /// Token id for each pseudo_output. This must have the same length as
+    /// `pseudo_output_commitments`, after mixed transactions feature.
+    pub pseudo_output_token_ids: Vec<u64>,
+
+    /// Token id for each output. This must have the same length as
+    /// `prefix.outputs`, after mixed transactions feature.
+    pub output_token_ids: Vec<u64>,
+}
+
+impl SigningData {
+    /// Create a new [`SigningData`] instance.
+    ///
+    /// # Arguments
+    /// * `block_version` - Block version of transaction being signed. This may
+    ///   influence details of the signature.
+    /// * `message` - The messages to be signed, e.g. Hash(TxPrefix).
+    /// * `rings` - Input rings, each one describing a single input to the
+    ///   transaction.
+    /// * `output_secrets` - Outputs secret (amount and commitment) for the
+    ///   outputs we will be creating.
+    /// * `fee` - Value of the implicit fee output.
+    /// * `check_value_is_preserved` - If true, check that the value of inputs
+    ///   equals value of outputs.
+    /// * `rng` - randomness
+    pub fn new<CSPRNG: RngCore + CryptoRng>(
+        block_version: BlockVersion,
+        message: &[u8; 32],
+        rings: &[InputRing],
+        output_secrets: &[OutputSecret],
+        fee: Amount,
+        check_value_is_preserved: bool,
+        rng: &mut CSPRNG,
+    ) -> Result<Self, Error> {
+        get_view_only_signing_data(
```

Comment:
I'd like to move the contents of `get_view_only_signing_data` but its a long method and will make the PR harder to review. I think it should be done in an immediate followup.